File: 2024/k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []
        for point in points:
            x = point[0]
            y = point[1]
            dist = (x*x) + (y*y)
            heap.append([dist, x, y])

        heapq.heapify(heap)
        res = []        
        # Find closest K points
        while k > 0:
            dist, x, y = heapq.heappop(heap)
            res.append([x,y])
            k -= 1
    
        return res

File: 2024-mk2/heaps-and-queues/k-closest-points-to-origin.go
 func kClosest(points [][]int, k int) [][]int {
    // For every point, calculate the distance
    // Store the distances in some sort of sortable data structure
    // We are using a minHeap here by using a priority queue that prioritizes smallest values being at the front. This way we can just iterate through k times and put them in the res.
    // Return the first K points
    
    // Create a maxHeap implementation with a priorityqueue
    maxHeap := priorityqueue.NewWith(func(a, b interface{}) int {
        distA := a.([]int)[2]
        distB := b.([]int)[2]
        if distA > distB {
            return 1
        } else if distA < distB {
            return -1
        }
        return 0
    })

    for _, p := range points {
        dist := p[0]*p[0] + p[1]*p[1]
        maxHeap.Enqueue([]int{p[0], p[1], dist})
    }
    
    res := [][]int{}
    for i:=0; i<k; i++ {
        val, _ := maxHeap.Dequeue()
        point := val.([]int)
        res = append(res, []int{point[0], point[1]})
    }


    return res
    
}
